### PR TITLE
add error check, status code great or equal than 500 would be in erro…

### PR DESCRIPTION
**Pull request by Jijun**
add error condition check , so we can filter error at skywalking ui. some java plugins  make error occured when status code  great or equal than 400, but at http side 4xx should be  "client error"
so i add  error tag when status code great or equal than 500

### CODE REVIEW

**Comment by wu-sheng** [comment]
Please recheck your codes, the e2e test fails.

**Comment by Jijun** [comment]
it's my fault, i forget to remove duplicate line about http.status tag.

**Comment by dmsolr** [comment]
It is basically good to me. But I wonder if this will lead to some misunderstanding.

**Comment by wu-sheng** [comment]
> It is basically good to me. But I wonder if this will lead to some misunderstanding.

What is your concern? Could you be clear?

**Comment by dmsolr** [comment]
In my mind, the Span ERROR means Nginx has some wrong or upstream returns an unsuccess(include 4xx and 5xx). I think that will make confusion to others.
Now, Tag is searchable. So, I think to make a new Tag to represent client_error(4xx), server_error(5xx). That will be better.

**Comment by wu-sheng** [comment]
tag=`status_code` is already available at the core level and other agents. We could add it here. 

Also, this logic is widely used in the Java agent too, take a look at this, https://github.com/apache/skywalking/blob/master/apm-sniffer/apm-sdk-plugin/httpClient-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/httpClient/v4/HttpClientExecuteInterceptor.java#L84-L87. You could find many codes like this. They even use `>=400`, rather than `>=500`.

**Comment by dmsolr** [comment]
Got it.

**Review by dmsolr — APPROVED** [review]
LGTM

File: lib/skywalking/tracer.lua [hunk 1]
```diff
@@ -68,6 +68,10 @@ end
 function Tracer:finish()
     -- Finish the exit span when received the first response package from upstream
     if ngx.ctx.exitSpan ~= nil then
+        local upstream_status = tonumber(ngx.var.upstream_status)
+        if upstream_status and upstream_status >= 500 then
+           Span.errorOccurred(ngx.ctx.exitSpan)
```
Comment (wu-sheng) [inline lib/skywalking/tracer.lua]: According to the discussion, I suggest you add `status_code` tag here.

Comment (wu-sheng) [reply lib/skywalking/tracer.lua]: @Jijun

File: lib/skywalking/tracer.lua [hunk 2]
```diff
@@ -77,7 +81,11 @@ function Tracer:prepareForReport()
     local TC = require('tracing_context')
     local Segment = require('segment')
     if ngx.ctx.entrySpan ~= nil then
-        Span.tag(ngx.ctx.entrySpan, 'http.status', ngx.var.status)
+        local ngxstatus = ngx.var.status
+        Span.tag(ngx.ctx.entrySpan, 'http.status', ngxstatus)
+        if tonumber(ngxstatus) >= 500 then
+           Span.errorOccurred(ngx.ctx.entrySpan)
+        end
```
Comment (wu-sheng) [inline lib/skywalking/tracer.lua:88]: Why do this again in the reporting stage? I think all logic should be included in `Tracer:finish`. Could you explain?

Comment (Jijun) [reply lib/skywalking/tracer.lua:88]: i follow the origin author's logic and made minimal changes as far as possible.

**Review by wu-sheng — APPROVED** [review]

**Review by wu-sheng — APPROVED** [review]

**Comment by wu-sheng** [comment]
@Jijun Please update main repo lua e2e with the latest master commit id?

**Pull request merged**